test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
+            delete: () => Promise.resolve(),
+          });
+        };
+        sessionPool.open();
+        return new Promise(resolve => setTimeout(resolve, 300)).then(() =>
+          sessionPool.getWriteSession().then(function(s) {
+            assert.strictEqual(s.type, 'readwrite');
+            sessionPool.release(s);
+          })
+        );
+      });
 
-      sessionPool.close();
+      it('should not get a read session as database is closed', function() {
+        sessionPool.session_ = function() {
+          return new Object({
+            create: () => Promise.reject(),
+            delete: () => Promise.resolve(),
+          });
+        };
+        return sessionPool.getReadSession().catch(error => {
+          assert.strictEqual(error.message, 'Database is closed.');
+        });
+      });
 
-      assert.strictEqual(destroyCallCount, fakeAll.length);
-      assert.deepEqual(sessionPool.reads_, []);
-      assert.deepEqual(sessionPool.writes_, []);
-      assert.deepEqual(sessionPool.borrowed_, []);
+      it('should not get a write session as database is closed', function() {
+        sessionPool.session_ = function() {
+          return new Object({
+            create: () => Promise.reject(),
+            delete: () => Promise.resolve(),
+          });
+        };
+        return sessionPool.getWriteSession().catch(error => {
+          assert.strictEqual(error.message, 'Database is closed.');
+        });
+      });
     });
 
-    it('should settle once all sessions are destroyed', function() {
-      var delay = 500;
-
-      sessionPool.reads_ = [{}];
-      sessionPool.destroySession_ = function() {
-        return new Promise(function(resolve) {
-          setTimeout(resolve, delay);
+    describe('release', function() {
+      it('verify that a read and write session are borrowed and then release them', function() {
+        sessionPool.createTransaction_ = () => Promise.resolve();
+        sessionPool.session_ = function() {
+          return new Object({
+            create: () => Promise.resolve(),
+            delete: () => Promise.resolve(),
+          });
+        };
+        sessionPool.open();
+        clearTimeout(sessionPool.pingTimeoutHandle);
+        return Promise.all([
+          sessionPool.getReadSession(),
+          sessionPool.getWriteSession(),
+        ]).then(function(sessions) {
+          const readSession = sessions[0];
+          const writeSession = sessions[1];
+          assert.strictEqual(sessionPool.getStats().writePool.borrowed, 1);
+          assert.strictEqual(sessionPool.getStats().readPool.borrowed, 1);
+          return Promise.all([
+            sessionPool.release(readSession),
+            sessionPool.release(writeSession),
+          ]);
         });
-      };
-
-      var end = timeSpan();
+      });
 
-      return sessionPool.close().then(function() {
-        assert(isAround(delay, end()));
+      it('should throw an error while release a write session', function() {
```

Comment:
Maybe rename this to something more descriptive like "should destroy the session when createTransaction fails"

Reply:
Fixed it